Validator UX docs for baklava network

aaitor opened the pull request:
### Description

As a validator I expected to have a valid documentation page allowing to me to run a validator in `baklava` network including the validator, proxy and attestation service.

### Tested

_An explanation of how the changes were tested or an explanation as to why they don't need to be._

### Other changes

_Describe any minor or "drive-by" changes here._

### Related issues

- Fixes #[issue number here]

### Backwards compatibility

_Brief explanation of why these changes are/are not backwards compatible._

aaitor committed:
Merge pull request #1 from celo-org/master

master

aaitor committed:
Merge pull request #2 from celo-org/master

master

aaitor committed:
Merge pull request #3 from celo-org/master

master

aaitor committed:
Merge pull request #4 from celo-org/master

master

aaitor committed:
Merge pull request #5 from celo-org/master

master

aaitor committed:
Merge pull request #6 from celo-org/master

master

aaitor committed:
Merge pull request #7 from celo-org/master

master

aaitor committed:
WIP: Adding baklava validator documentation

jcortejoso committed:
Added command for getting proxy ip

aaitor committed:
doc editing

aaitor committed:
last edits

aaitor committed:
Merge branch 'feature/baklava_validator_doc' of github.com:aaitor/celo-monorepo into feature/baklava_validator_doc

aaitor committed:
adapting to new proxy and validator commands

aaitor committed:
toc update & alfajores commands review & baklava rewriting

aaitor commented:
This PR **shouldn't be merged** till the official baklava images are publicly available and all the commands are tested again using those images.

aaitor committed:
adding attestation service initial documentation

jcortejoso committed:
Minnor changes. Added missing option to proxy

jcortejoso commented on packages/docs/getting-started/running-a-validator.md:
```diff
@@ -48,67 +64,258 @@ A $ signifies the bash prompt. Everything following it is the command you should
+
+{% hint style="danger" %}
+**Security**: The command line above includes the parameter `--rpcaddr 0.0.0.0` which makes the Celo Blockchain software listen for incoming RPC requests on all network adaptors. Exercise extreme caution in doing this when running outside Docker, as it means that any unlocked accounts and their funds may be accessed from other machines on the Internet. In the context of running a Docker container ...
+{% endhint %}
```
So I would say this is not exactly correct although the warning can be good to have. The flag `--rpcaddr 0.0.0.0` does not have a real implication bc 0.0.0.0 is referring to all the interfaces of the docker container, that will be isolated from the host interfaces (the are connected by a virtual network bridge).
The control of the host interfaces listening are in docker option `-p` or `--publish`, so we can adapt to that.

aaitor replied:
slightly adapted

jcortejoso committed:
Merge branch 'master' into feature/baklava_validator_doc

aaitor committed:
removing `miner.verificationpool` flag for baklava validators

aaitor committed:
Merge branch 'feature/baklava_validator_doc' of github.com:aaitor/celo-monorepo into feature/baklava_validator_doc

aaitor committed:
added documentation about how to run a full node in baklava network

aaitor committed:
fixing command

jcortejoso committed:
Added draft for env variables table

jcortejoso committed:
Merge branch 'feature/baklava_validator_doc' of github.com:aaitor/celo-monorepo into feature/baklava_validator_doc

aaitor committed:
Merge branch 'feature/baklava_validator_doc' of github.com:aaitor/celo-monorepo into feature/baklava_validator_doc

aaitor committed:
changing reference to baklava faucet

aaitor committed:
adding commands

aaitor committed:
Merge pull request #8 from celo-org/master

master

aaitor committed:
Merge branch 'master' into feature/baklava_validator_doc

aaitor committed:
fixing bash commands

aaitor committed:
adding some commands

tkporter commented on packages/docs/getting-started/running-a-validator.md:
```diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
-Run:
+To run a complete validator it's necessary to execute the following components:
+
+- The valitor software
```
nit: validator

tkporter commented on packages/docs/getting-started/running-a-validator.md:
```diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+### Environment variables
 
-We're going to use a Docker image containing the Celo node software in this tutorial.
+| Variable                     | Explanation                                                                   | Default Value |
```
should there be default values?

aaitor replied:
removing useless column

tkporter commented on packages/docs/getting-started/running-a-validator.md:
```diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
-We're going to use a Docker image containing the Celo node software in this tutorial.
+| Variable                     | Explanation                                                                   | Default Value |
+| ---------------------------- | ----------------------------------------------------------------------------- | ------------- |
+| CELO_IMAGE                   | The docker image used for the validator and proxy containers                  |               |
```
nit: docker -> Docker

aaitor replied:
:+1:

tkporter commented on packages/docs/getting-started/running-a-validator.md:
```diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+| ---------------------------- | ----------------------------------------------------------------------------- | ------------- |
+| CELO_IMAGE                   | The docker image used for the validator and proxy containers                  |               |
+| CELO_NETWORK                 | The Celo network to connect with. This variable is also used as the image tag |               |
+| NETWORK_ID                   | The celo network chain id                                                     |               |
```
nit: celo -> Celo, maybe id -> ID

aaitor replied:
:+1:

tkporter commented on packages/docs/getting-started/running-a-validator.md:
```diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+| CELO_IMAGE                   | The docker image used for the validator and proxy containers                  |               |
+| CELO_NETWORK                 | The Celo network to connect with. This variable is also used as the image tag |               |
+| NETWORK_ID                   | The celo network chain id                                                     |               |
+| URL_VERIFICATION_POOL        | URL for the Verification pool for the attestation process                     |               |
```
I don't think this is needed anymore now that we have the attestation service

aaitor replied:
removed from `baklava` documentation

tkporter commented on packages/docs/getting-started/running-a-validator.md:
```diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+To run a complete validator it's necessary to execute the following components:
+
+- The valitor software
+- A Proxy that acts as an intermediary for the validator requests
```
Wondering if we should make it clearer that a proxy isn't necessarily required & maybe a quick sentence about the rationale

aaitor replied:
makes sense, adding a sentence to clarify

tkporter commented on packages/docs/getting-started/running-a-validator.md:
```diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
 
+We are going to need to create 3 accounts, 2 for the validator and 1 for the Proxy.
+
+First we create three accounts, one for the Validator, one for Validator Group and the last one for the Proxy. You can get their addresses if you don’t already have them. If you already have some accounts, you can skip this step.
```
"one for Validator Group" -> "one for the Validator Group"

Maybe make it clearer that "get their addresses" is in reference to later in this step-- "generate their addresses in this step" or something

aaitor replied:
rephrased

tkporter commented on packages/docs/getting-started/running-a-validator.md:
```diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+
+#### Deploy the validator and proxy nodes
+
+We initialize the docker containers for the validator and the proxy, building from an image for the network and initializing Celo with the genesis block:
```
I could see people being confused where `/celo/genesis.json` came from, so maybe specify "with the genesis block found inside the Docker image"

aaitor replied:
changed

tkporter commented on packages/docs/getting-started/running-a-validator.md:
````diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+docker run -v $PWD/proxy:/root/.celo --entrypoint /bin/sh -it $CELO_IMAGE:$CELO_NETWORK -c "geth account new"
+```
+
+Those commands will prompt you for a passphrase, ask you to confirm it, and then will output your account address: `Address: {<YOUR-ACCOUNT-ADDRESS>}`
````
I believe in the past we've had issues with the prompt, might be worth adding a note about it being a known issue? See first part here https://forum.celo.org/t/setting-up-a-validator-faq/90

aaitor replied:
adding a `warning` message

tkporter commented on packages/docs/getting-started/running-a-validator.md:
````diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+docker run -v $PWD/validator:/root/.celo $CELO_IMAGE:$CELO_NETWORK init /celo/genesis.json
+```
+
+To participate in consensus, we need to set up our nodekey for our accounts. We can do so via the following commands \(it will prompt you for your passphrase\):
````
Same thing about prompt issues

aaitor replied:
:+1:

tkporter commented on packages/docs/getting-started/running-a-validator.md:
````diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+At this point we are ready to start up the proxy:
+
+```bash
+docker run --name celo-proxy -p 8545:8545 -p 8546:8546 -p 30303:30303 -p 30303:30303/udp -p 30503:30503 -p 30503:30503/udp -v $PWD/proxy:/root/.celo $CELO_IMAGE:$CELO_NETWORK --verbosity 3 --networkid $NETWORK_ID --syncmode full --rpc --rpcaddr 0.0.0.0 --rpcapi eth,net,web3,debug --maxpeers 1100 --etherbase=$CELO_PROXY_ADDRESS --proxy.proxy --proxy.proxiedvalidatoraddress $CELO_VALIDATOR_ADDRESS  ...
````
We should add `--restart=Always` so that if the docker container crashes (a real possibility due to the memleak in geth) it'll be automatically restarted

Also, this will run it in the foreground, a note about using `-d` to run in background could be useful to those unfamiliar with Docker. But then they won't be able to see logs as easily, so I'm not super passionate

aaitor replied:
:+1:

tkporter commented on packages/docs/getting-started/running-a-validator.md:
````diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+Now we can start up the validator node:
+
+```bash
+docker run --name celo-validator -p 127.0.0.1:8547:8545 -p 127.0.0.1:8548:8546 -p 30304:30303 -p 30304:30303/udp -v $PWD/validator:/root/.celo $CELO_IMAGE:$CELO_NETWORK --verbosity 3 --networkid $NETWORK_ID --syncmode full --rpc --rpcaddr 0.0.0.0 --rpcapi eth,net,web3,debug --maxpeers 125 --mine --istanbul.blockperiod=1 --istanbul.requesttimeout=3000 --etherbase $CELO_VALIDATOR_ADDRESS --nodiscov ...
````
--istanbul.blockperiod should be 5, same note about `--restart=Always`

aaitor replied:
:+1:

tkporter commented on packages/docs/getting-started/running-a-validator.md:
````diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+celocli account:new
+```
+
+We copy the account details and assign the Private Key to the `ATTESTATION_SERVICE` environment variable:
````
`ATTESTATION_SERVICE` -> `ATTESTATION_KEY`

aaitor replied:
:+1:

tkporter commented on packages/docs/getting-started/running-a-validator.md:
````diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+Add your URL:
+
+```bash
+celocli identity:change-attestation-service-url ./metadata.json --url ATTESTATION_SERVICE_URL
````
I think it should be clearer what `ATTESTATION_SERVICE_URL` is (also should it have a `$` or <>?)

aaitor replied:
:+1:

tkporter commented on packages/docs/getting-started/running-a-validator.md:
````diff
@@ -46,33 +66,227 @@ The code you'll see on this page is bash commands and their output.
+celocli identity:change-attestation-service-url ./metadata.json --url ATTESTATION_SERVICE_URL
+```
+
+And then host your metadata somewhere reachable via HTTP. You can register your metadata URL with:
````
might be a good idea to suggest a gist and to use the url to the raw text

aaitor replied:
:+1:

tkporter reviewed (commented):
Nice stuff! Some general comments:
* Mixed capitalization of Proxy/proxy, Validator/validator, ID/id, IP/ip, Docker/docker-- best to choose one and stick with it
* I think we should make it more clear what the purpose of a proxy node is and that we would want a proxy node and a proxied validator to be run ideally on two separate machines

aaitor committed:
separating running a validator page in alfajores and baklava pages

aaitor committed:
Merge pull request #9 from celo-org/master

master

aaitor committed:
Merge branch 'master' into feature/baklava_validator_doc

aaitor committed:
adding configuration steps about attestation service

aaitor committed:
database section

aaitor committed:
adding yarn script to support `yarn run db:create` command

aaitor committed:
moving attestation service doc to independent page and linking from baklava and alfajores pages

aaitor committed:
fixing comments in PR

aaitor committed:
unifying case setup for proxy, validator and attestation service

aaitor committed:
Merge branch 'master' into feature/baklava_validator_doc

aaitor committed:
small clarification of security warning

aaitor committed:
Adding a script for running a docker network with a validator and a proxy.
The intention is to add also the attestation-service, allowing to replicate
and automate the executiong of the running-a-validator.md guide

aaitor committed:
Merge branch 'feature/baklava_validator_doc' of github.com:aaitor/celo-monorepo into feature/baklava_validator_doc

aaitor committed:
removing CELO_NETWORK variable for simplicity

aaitor committed:
fixing deploy block

aaitor committed:
adding screen to the loop allowing to attach/deattach to multiple sessions

aaitor committed:
removing commented commands

aaitor committed:
Adding attestation service to the docker test script.
Needs to be tested with a publicly available attestation-service docker image as soon as this as published.

aaitor committed:
specifying the right attestation service web3 provider config

aaitor committed:
fixing the validator account unlock

aaitor committed:
changing attestation service starting command

aaitor committed:
printing validator and proxy status messages

aaitor committed:
fixing quotes

aaitor committed:
adding reference to running a validator script

nambrot commented on packages/attestation-service/package.json:
```diff
@@ -19,6 +19,7 @@
     "prepublishOnly": "yarn build:gen && yarn build",
     "start-ts": "TS_NODE_FILES=true ts-node src/index.ts",
     "start": "node lib/index.js",
+    "db:create": "mkdir -p db && touch db/dev.db",
```
I don't think this should be here? I.e. it should not be the concern of this application to create databases other than for development and testing

aaitor replied:
that command was to make it consistent with the documentation found in the README.md of the attestation-service and the comment made about not including the `dev` suffix.

What do you suggest? Add the command or keep the `yarn run db:create:dev` instruction in the README?

nambrot commented on packages/docs/getting-started/running-a-validator-baklava.md:
```diff
@@ -0,0 +1,178 @@
+| CELO_VALIDATOR_POP           |                                                                           |  |
+| PROXY_ENODE                  | The ethereum node address for the Validator                               |  |
+| PROXY_IP                     | The Proxy container internal IP address from docker pool address          |  |
+| ATTESTATION_KEY              | The etherbase private key for the account used in the Attestation Service |  |
```
Not sure if `etherbase private key` is the best description for this key?

nambrot commented on packages/docs/getting-started/running-a-validator-baklava.md:
````diff
@@ -0,0 +1,178 @@
+docker pull $CELO_IMAGE
+```
+
+### Create accounts
````
IMO this section should cover the creation of the attestation key

aaitor replied:
To avoid duplicate this documentation in two pages, I added the text clarifying are necessary 4 accounts (Validator, Validator Group, Proxy and Attestation Service) and added a link to the attestation service documentation page.

nambrot commented on packages/docs/getting-started/running-attestation-service.md:
```diff
@@ -0,0 +1,125 @@
+| DATABASE_URL    | The URL under which your database is accessible, currently supported are `postgres://`, `mysql://` and `sqlite://`                                                     |  |
+| CELO_PROVIDER   | The URL under which a celo blockchain node is reachable, i.e. something like `https://integration-forno.celo-testnet.org`                                              |  |
+| ACCOUNT_ADDRESS | The address of the validator account                                                                                                                                   |  |
+| ATTESTATION_KEY | The private key with which attestations should be signed. You could use your account key for attestations, but really you should authorize a dedicated attestation key |  |
```
We should probably just tell people to delegate an attestation signer, it's actually quite hard to retrieve the account key for use anyways

nambrot commented on packages/docs/getting-started/running-attestation-service.md:
````diff
@@ -0,0 +1,125 @@
+celocli account:new
+```
+
+We copy the account details and assign the Private Key to the `ATTESTATION_KEY` environment variable:
````
We need to authorize the key on the `Accounts` smart contract

nambrot commented on packages/docs/getting-started/running-attestation-service.md:
````diff
@@ -0,0 +1,125 @@
+You can find the migration scripts for creating the schema at the `celo-monorepo`, `packages/attestation-service` folder. From there, after setting up the `DATABASE_URL` env variable you can run the following commands:
+
+```bash
+yarn run db:create
````
This won't work other than the `sqlite` driver

aaitor replied:
removing that `db:create` line and adding reference that depending on the database technology the users need to create the database with the proper user access.

nambrot reviewed (approved):
Great stuff! There are some inaccuracies, but I'm happy to merge and improve in follow-on PRs

jcortejoso committed:
Merge branch 'master' into feature/baklava_validator_doc

nambrot committed:
Merge branch 'master' into feature/baklava_validator_doc
